modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+  For more constants that can used alongside refer :mod:`Sys`.
+
+  To those familiar with the Unix socket API, the method names will feel familiar,
+  though their usage will be somewhat simpler than the raw Unix socket API.
+
+  Records
+  -----------------
+  :type:`ipAddr`
+  :type:`tcpConn`
+  :type:`tcpListener`
+  :type:`udpSocket`
+
+  Enum
+  -----------------
+  :type:`IPFamily`
+
+  Procedures
+  -----------------
+  :proc:`bind`
+  :proc:`connect`
+  :proc:`delayAck`
+  :proc:`getpeername`
+  :proc:`getSockOpt`
+  :proc:`getsockname`
+  :proc:`listen`
+  :proc:`naggle`
+  :proc:`setSockOpt`
+
+  Records, Types and Function Definitions
+  ---------------------------------------
+*/
+module Socket {
+
+// TODO: replace with mason config
+require "/usr/include/event2/event.h";
+require "/usr/include/event2/thread.h";
+require "-levent";
+require "-levent_pthreads";
+
+public use Sys;
+public use SysError;
+use Time;
+use SysCTypes;
+use SysBasic;
+use CPtr;
+use IO;
+
+/*
+  Available values for different Internet
+  Protocol Families to be used when creating Sockets.
+*/
+enum IPFamily {
+  /* IPv4 */
+  IPv4 = 2,
+  /* IPv6 */
+  IPv6 = 10,
+  /* IP Unspecified */
+  IPUnspec = 0
+}
+
+/* Type of Standard IPv4 Address */
+type ipv4Addr = sys_in_addr_t;
+/* Standard IPv4 Addresses of type :type:`ipv4Addr` */
+const IPv4Localhost:ipv4Addr = INADDR_LOOPBACK;
+const IPv4Any:ipv4Addr = INADDR_ANY;
+const IPv4Broadcast:ipv4Addr = INADDR_BROADCAST;
+
+/* Type of Standard IPv6 Address */
+type ipv6Addr = sys_in6_addr_t;
+/* Standard IPv6 Addresses of type :type:`ipv6Addr` */
+const IPv6Localhost:ipv6Addr = in6addr_loopback;
+const IPv6Any:ipv6Addr = in6addr_any;
+
+pragma "no doc"
+proc sys_sockaddr_t.init(in other: sys_sockaddr_t) {
+  this.init();
+  try! {
+    var host = other.numericHost();
+    var port = other.port();
+    var family = other.family;
+    this.set(host.c_str(), port, family:c_int);
+  }
+}
+
+/*
+  Abstract supertype for network addresses. Contains data
+  about :type:`IPFamily`, `host` and `port`.
+*/
+record ipAddr {
+  pragma "no doc"
+  var _addressStorage:sys_sockaddr_t;
+}
+
+pragma "no doc"
+proc ipAddr.init(in address: sys_sockaddr_t) {
+  this._addressStorage = new sys_sockaddr_t(address);
+}
+
+/*
+  Returns a new record of type :type:`ipAddr` prvoided `host`, `port`
+  and `family`. this function is equivalent to the following code:
+
+  .. code-block:: Chapel
+
+    ipAddr.create("127.0.0.1", 8111, IPFamily.IPv4)
+
+  :arg host: address string in dot-dash or colon notation depending on family.
+  :type host: `string`
+  :arg port: network address's port.
+  :type port: `uint(16)`
+  :arg family: value of IP Family
+  :type family: :type:`IPFamily`
+  :return: address instance.
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.create(host: string = "127.0.0.1", port: uint(16) = 8000, family: IPFamily = IPFamily.IPv4): ipAddr throws {
```

Comment:
I was expecting one to write `new ipAddr("127.0.0.1", 8111, IPFamily.IPv4)` etc. See https://github.com/chapel-lang/chapel/pull/17899#issuecomment-923007197

Reply:
Added the comment in design doc

Reply:
Can you add a comment here as well please? (It could be inside of the function - it is just for maintainers to keep track)